### PR TITLE
[GPU] Add support for linear tree with device=gpu

### PR DESCRIPTION
**Pull request by dragonbra**
GPU version's implementation refer to issue #6555

### CODE REVIEW

**Comment by dragonbra** [comment]
@microsoft-github-policy-service agree

**Review by StrikerRUS — COMMENTED** [review]
@dragonbra Thanks a lot for your contribution!

I checked that there are no any other differences except "gpu_" prefixes in new files compared to original `linear_tree_learner.h` and `linear_tree_learner.cpp`.

I don't think that my review should be count to allow the merge. But I left one wording suggestion and would like to ask you add new files in `LightGBM.vcxproj` and `LightGBM.vcxproj.filters`: https://github.com/microsoft/LightGBM/tree/master/windows.

File: src/io/config.cpp [hunk 1]
```diff
@@ -417,9 +417,9 @@ void Config::CheckParamConflict(const std::unordered_map<std::string, std::strin
   }
   // linear tree learner must be serial type and run on CPU device
   if (linear_tree) {
-    if (device_type != std::string("cpu")) {
+    if (device_type != std::string("cpu") && device_type != std::string("gpu")) {
       device_type = "cpu";
-      Log::Warning("Linear tree learner only works with CPU.");
+      Log::Warning("Linear tree learner only works with CPU and GPU.");
```
[the remark this answers is not in the document]
Comment (dragonbra) [reply src/io/config.cpp:422]: Thank you, I will take this wording advice!

**Comment by dragonbra** [comment]
> @dragonbra Thanks a lot for your contribution!
> 
> I checked that there are no any other differences except "gpu_" prefixes in new files compared to original `linear_tree_learner.h` and `linear_tree_learner.cpp`.
> 
> I don't think that my review should be count to allow the merge. But I left one wording suggestion and would like to ask you add new files in `LightGBM.vcxproj` and `LightGBM.vcxproj.filters`: https://github.com/microsoft/LightGBM/tree/master/windows.

Hi, I checked the files `LightGBM.vcxproj` and `LightGBM.vcxproj.filters` but I didn't see `gpu_tree_learner.h` in them. I wonder if I should add things like `gpu_linear_tree_learner.h` in them without `gpu_tree_learner.h`.

**Comment by StrikerRUS** [comment]
@dragonbra Thanks for checking this! Indeed, there are no any `gpu_*` files. So I think we shouldn't add your new files there as well.

File: src/treelearner/gpu_linear_tree_learner.cpp [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+ */
+#include "gpu_linear_tree_learner.h"
+
+#include <Eigen/Dense>
+
+#include <algorithm>
+
+namespace LightGBM {
+
+void GPULinearTreeLearner::Init(const Dataset* train_data, bool is_constant_hessian) {
```
Comment (shiyu1994) [inline src/treelearner/gpu_linear_tree_learner.cpp:13]: Overall, the solution looks good to me. I've read half of the implementation and need some more time to check the correctness. Thank you.

For test case, I think the test case for this scenario is automatically enabled with existing test cases for linear trees and `device_type=gpu`.

**Review by shiyu1994 — APPROVED** [review]

**Comment by shiyu1994** [comment]
Given the redundancy in `gpu_linear_tree_learner.cpp` and `linear_tree_learner.cpp`, I will try to merge these two files. Will push some changes to remove the duplicate code soon.

**Review by shiyu1994 — APPROVED** [review]

**Comment by jameslamb** [comment]
Thank you very much for simplifying this to reduce the code duplication @shiyu1994 !

**Pull request merged**